### PR TITLE
feat: wire up persistence in ingester

### PR DESCRIPTION
**Pull request by pauldix**
This is functional with the happy path test in place. There are a number of followup pieces to complete ingester persistence, which I've put into a list in #3705.

For now I've opted to panic on unexpected state so that this gets surfaced while we're in test and validation mode. We'll likely want to change this to log those issues and alert operators to perform replay and recovery on specific Namespaces, allowing everything else to continue to operate.

### CODE REVIEW

File: ingester/src/data.rs [hunk 1]
```diff
@@ -511,6 +675,49 @@ impl DataBuffer {
...
+    /// if there is already a persisting batch.
+    pub fn snapshot_to_persisting(
+        &mut self,
+        sequencer_id: SequencerId,
+        table_id: TableId,
+        partition_id: PartitionId,
+        table_name: &str,
+    ) -> Arc<PersistingBatch> {
+        if self.persisting.is_some() {
+            panic!("can't snapsoht while already persisting (handle this)")
```
Comment (e-dard) [inline ingester/src/data.rs:693]: typo but I guess this is getting removed before merge.

**Review by e-dard — COMMENTED** [review]
Though this PR is still in-flight I took a look through the code and it LGTM 👍

File: ingester/src/data.rs [hunk 2]
```diff
@@ -110,13 +123,119 @@ pub(crate) trait Persister: Send + Sync + 'static {
...
+        };
+
+        let partition_data =
+            match table_data.partition_data(&partition_info.partition.partition_key) {
+                None => {
+                    error!(%partition_info.partition.id, "partition not in ingester state");
+                    return;
+                }
+                Some(p) => p,
+            };
```
Comment (domodwyer) [inline ingester/src/data.rs:180]: It makes me uneasy that the caller has no way of knowing if the persist operations worked or not (even if the caller decides not to act on any failures today). Perhaps this is my personal design bias though.

If you had an error type, this block could be refactored down to something nice like:

```suggestion
        let partition_info = partition_info.ok_or(PersistError::PartitionNotFound)?;
        let sequencer_data = self.sequencers.get(&partition_info.partition.sequencer_id).ok_or(PersistError::SequencerNotFound)?;
        let namespace = sequencer_data.namespace(&partition_info.namespace_name).ok_or(PersistError::NamespaceNotFound)?;
        let table_data = namespace.table_data(&partition_info.table_name).ok_or(PersistError::TableNotFound)?;
        let partition_data = table_data.partition_data(&partition_info.partition.partition_key).ok_or(PersistError::PartitionNotFound)?;
```

And let the caller decide what to do/log/etc - this would help with retries, this is a pattern I use a lot:

```rust
/// The pub persist method
pub fn persist(...) -> Result<()> {
    loop {
        /// Calls an non-pub method that tries to persist once and returns a descriptive error type
        match persist_inner(...) {
            Ok(_) => return Ok(()),
            Err(PersistError::PartitionNotFound) => {
                // log, maybe continue the loop to retry, or return, etc
            }
            Err(PersistError::SequencerNotFound) => {
                // log, maybe continue the loop to retry, or return, etc
            }
        }

        // Maybe sleep / back off to prevent a hot retry loop
        tokio::time::sleep(1s)
    }
}
```

**Review by domodwyer — COMMENTED** [review]
I was looking through the code this PR interacts with and noticed a bug in the lifecycle manager:

```rust
    pub async fn maybe_persist<P: Persister>(&self, persister: &Arc<P>) {
        // ensure that this is only running one at a time
        self.persist_running.lock().await;

        // ...
    }
```

Here the `lock()` method returns a [`MutexGuard`](https://docs.rs/tokio/1.16.1/tokio/sync/struct.MutexGuard.html) that is documented as releasing the lock when dropped:

```
As long as you have this guard, you have exclusive access to the underlying T. 

...

The lock is automatically released whenever the guard is dropped, at which point lock will succeed yet again.
```

However because we're not taking the guard (no `let x = lock()`) it is dropped as soon as it is acquired - see [this for an example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=435136ea73b310bb795f2891d5da7c84).

If we change the lock line to be:

```rust
        let _guard = self.persist_running.lock().await;
```

We hold onto the guard for the entire `maybe_persist()` call and therefore the lock will be held for the entire call which I think is the intention.

File: ingester/src/data.rs [hunk 3]
```diff
@@ -110,13 +123,119 @@ pub(crate) trait Persister: Send + Sync + 'static {
...
+        let partition_info = {
+            let mut txn = self
+                .catalog
+                .start_transaction()
+                .await
+                .expect("make this retry");
+            txn.partitions()
+                .partition_info_by_id(partition_id)
+                .await
+                .expect("make this retry")
```
[the remark this answers is not in the document]
Comment (pauldix) [reply ingester/src/data.rs:137]: Yeah, I want to update the Catalog API to not require transactions at all for requests that just want to do a simple select, which is actually most requests to the catalog. The new API was definitely a pain point while doing this persist code. Also, I have a concern that using a two-phase transaction where a simple prepared statement will do will put extra load on the Postgres server, but I haven't had time to actually research if this fear has any basis in reality.

[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply ingester/src/data.rs:137]: I just reach here and have not read the rest but here is my opinion. We only need to call `start_transaction` if we want to to write/update/delete something. A `start_transaction` always go with a `commit` or `rollback` at the end. If we just need to read something, we do no need start_transaction if we only read data before certain sequence_number

File: ingester/src/data.rs (hunk 2 above)
Comment (pauldix) [reply ingester/src/data.rs:180]: I definitely had the same concern. Although the pattern you list has the problem that it retries the entire persist operation, where I actually only want to retry on the thing that failed and continue on from there.

Thinking about the errors there are two situations (I think):
1. Errors related to some unexpected state
2. Errors related to the Catalog or the Object Store being unavailable or too busy

For errors in the second category, we just want to retry forever until it succeeds. And we don't want to repeat all the work that we did to get to that point. So if we fail on the last step to update the Catalog, we should just keep retrying, rather than redoing the snapshot, compaction, object store write, etc.

For errors in the first category, I'm not quite sure what to do. The state that created the error is unexpected. Either someone else modified the catalog underneath us (dropping a partition for example) or the buffered state in memory is off (some other caller dropped something (currently not possible), or the mutable batch error (which I don't think should actually be possible).

Basically, what do we do in the case of unexpected state? Log the problem, evict the data and move on? Log the problem and continue backing up until the system pauses ingest? Log and crash?

File: ingester/src/data.rs [hunk 4]
```diff
@@ -110,13 +123,119 @@ pub(crate) trait Persister: Send + Sync + 'static {
...
+                .await
+                .expect("make this retry");
+
+            add_parquet_file_with_tombstones(
+                &parquet_file,
+                &persisting_batch.data.deletes,
+                txn.deref_mut(),
+            )
+            .await
+            .expect("make this retry");
```
Comment (NGA-TRAN) [inline ingester/src/data.rs:230]: You need to call `rollback` if this fails because the failure may happen at inserting processed tombstones after the parquet file already inserted.

Comment (NGA-TRAN) [reply ingester/src/data.rs:230]: I think better way to handle this is the `start _transaction`, `rollback` and `commit` should be done inside  `add_parquet_file_with_tombstones`. This caller just needs to verify success or failure.

File: ingester/src/data.rs [hunk 5]
```diff
@@ -269,6 +388,31 @@ impl NamespaceData {
...
+
+        if let Some(t) = table {
+            let mut partitions = t.partition_data.write();
+            let partition = partitions.get(partition_key).cloned();
+
+            if let Some(p) = partition {
+                let mut data = p.inner.write();
+                data.persisting = None;
+                if data.is_empty() {
+                    partitions.remove(partition_key);
```
Comment (NGA-TRAN) [inline ingester/src/data.rs:407]: I am not sure if you need a lock here

File: ingester/src/data.rs [hunk 6]
```diff
@@ -269,6 +388,31 @@ impl NamespaceData {
...
+                }
+            }
+
+            if partitions.is_empty() {
+                tables.remove(table_name);
```
Comment (NGA-TRAN) [inline ingester/src/data.rs:412]: same here

File: ingester/src/data.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply ingester/src/data.rs:137]: Just went over the code, we do not need start_transaction here. The only transaction needed  should be inside `add_parquet_file_with_tombstones'. I think @crepererum modified it when he was adding the abstraction layer. Let me have a look and see if I can put it back in

File: ingester/src/data.rs (hunk 2 above)
Comment (mkmik) [reply ingester/src/data.rs:180]: > Basically, what do we do in the case of unexpected state? 

let-it-crash would be a safe bet, but I'm not sure what's the right tradeoff here with the ingestors acting as an in-memory database for a portion of the data.

> For errors in the second category, we just want to retry forever until it succeeds

It think that's a reasonable approach. There are some edge cases like the error not being a transient one but being something that depends on the file you're writing (e.g. path length errors, or path encoding errors, or just unlucky assignments to some shard on the object storage upstream servers). If you're too eager to retry a set of s3 objects that will always fail despite most of the other ones succeeding, you may cause an unnecessarily wide outage. But that being said, it's probably not worth it being too clever at this stage.

File: ingester/src/data.rs (hunk 5 above)
Comment (pauldix) [reply ingester/src/data.rs:407]: Need the lock to set the persisting to None and also need to keep it while we check to see if we want to remove the partition, same with the table lock. Because the ingester might be pulling in data from Kafka to the same partition at the same time so we want to hold a lock across the entire namespace while we clear this state.

Comment (NGA-TRAN) [reply ingester/src/data.rs:407]: Oh sorry, I did not see the `write` lock above. What I meant was "why do you not take lock here?" which you did :)

**Review by NGA-TRAN — APPROVED** [review]
Since you already discussed the transaction changes  and future error  handling. I approve this for Paul to move forward after listing all needed todos

File: ingester/src/data.rs (hunk 2 above)
Comment (pauldix) [reply ingester/src/data.rs]: I've added the retry loops. On the state machine, I'll have to reread that post (I saw it quite a while ago). But to be honest, I found the code significantly harder to follow than a simple read of top to bottom down a function.

File: ingester/src/data.rs [hunk 7]
```diff
@@ -111,13 +124,155 @@ pub(crate) trait Persister: Send + Sync + 'static {
...
+        // snapshot and make arc clones of the data.
+        let persisting_batch = partition_data.snapshot_to_persisting_batch(
+            partition_info.partition.sequencer_id,
+            partition_info.partition.table_id,
+            partition_info.partition.id,
+            &partition_info.table_name,
+        );
+
+        // do the CPU intensive work of compaction, de-duplication and sorting
+        let exec = Executor::new(1);
```
Comment (crepererum) [inline ingester/src/data.rs:199]: The executor should be global and cached for the following reasons:

- creating executors is potentially expensive
- the number of parallel "persist" tasks is not limited by the runtime, so you're creating an unbound number of threads here, at best overloading the server (we normally try to create about as many threads as there are logical cores assigned to the container), at worst you're running out of file descriptors killing the entire ingester process

You may just initialize the global executor w/ a single thread for now and we can do the "how many threads" config wiring from CLI (+ auto-detection of the default value) later.

File: ingester/src/data.rs [hunk 8]
```diff
@@ -111,13 +124,155 @@ pub(crate) trait Persister: Send + Sync + 'static {
...
 #[async_trait]
 impl Persister for IngesterData {
-    async fn persist(&self, _partition_id: PartitionId) {
-        // lookup the TableData
-        // let persisting_batch = table_data.create_persisting_batch(partition.partition_key);
-        // do the persist with this persisting batch
-        // update the catalog
-        // table_data.clear_persisting_batch() (behind the scenes this will remove the persisting batch
-        // and if the partition is empty, remove it from the map in table_data)
+    async fn persist(&self, partition_id: PartitionId) {
```
Comment (crepererum) [inline ingester/src/data.rs:127]: What happens if you call this function w/ the same `partition_id` twice, precisely:

- 2nd call starts after 1st one finishes
- 2nd call starts while 1st one is in progress

I don't see any technical measure that you prevent that.

**Review by crepererum — COMMENTED** [review]
I'm a bit uneasy with the overall locking strategy. It seems that basically all objects in the ingester have interior mutability. While is this also true for the old DB catalog, we have clear lifecycle markers there that prevent you from touching bits and pieces while they're "in-flight", e.g. you cannot write to a "chunk" (here it would be a partition) while persistence is running. I don't see any of this in the new system and that just asks for heisenbugs.

**Comment by pauldix** [comment]
Responding to this part of @crepererum's comment:
> we have clear lifecycle markers there that prevent you from touching bits and pieces while they're "in-flight", e.g. you cannot write to a "chunk" (here it would be a partition) while persistence is running.
A partition buffer can always be written to whether persistence is running or not.

File: ingester/src/data.rs (hunk 8 above)
Comment (pauldix) [reply ingester/src/data.rs:127]: It panics when you call `partition_data.snapshot_to_persisting_batch`. Happens on line 730. Enforced that for a given partition, only one persist can be running at any given time. Although many partitions may be persisting at the same time.

Comment (crepererum) [reply ingester/src/data.rs:127]: This can be done in a follow-up PR, but I would like to seed a test that triggers this panic to make sure this guard / invariant holds even when we refactor stuff.

**Comment by crepererum** [comment]
> A partition buffer can always be written to whether persistence is running or not. When persistence is triggered, the buffer in the partition gets snapshotted and everything gets moved into the `PersistingBatch` (this happens while holding a lock on the partition's buffer). **After that snapshot, data can be written into the buffer while that `PersistingBatch` is compacted (deletes, duplicates, sorting) and then written into a parquet file in object storage.**

Makes sense. I would like to see some test for the highlighted scenario (in a later PR).

File: ingester/src/data.rs (hunk 8 above)
Comment (pauldix) [reply ingester/src/data.rs:142]: Ok, I've added #3755 to track adding that test

**Comment by pauldix** [comment]
Ok, I believe this is ready to go. I add a bit to log the writes into the lifecycle manager and wire up the pause for ingest.

File: ingester/src/data.rs [hunk 9]
```diff
@@ -79,48 +93,197 @@ pub struct IngesterData {
...
+                }
+            }
+        }
+
+        // Commit the parquet file and tombstones to the catalog. This is pretty ugly because of all
+        // the failures that might happen where we just want to keep retrying it.
+        // TODO: clean this up when updating the min_sequence_number is added in.
+        let parquet_file = iox_meta.to_parquet_file();
+        loop {
+            match self.catalog.start_transaction().await {
```
Comment (NGA-TRAN) [inline ingester/src/data.rs:250]: It is fine if you want to keep the start and commit transaction here but I think it will be cleaner if you start and commit the transaction inside the iox_catalog::add_parquet_file_with_tombstone. I am fine with either and I can also help move them inside the function later, too.

File: ingester/src/data.rs [hunk 10]
```diff
@@ -79,48 +93,197 @@ pub struct IngesterData {
...
+            partition_info.partition.sequencer_id,
+            partition_info.partition.table_id,
+            partition_info.partition.id,
+            &partition_info.table_name,
+        );
+
+        // do the CPU intensive work of compaction, de-duplication and sorting
+        let (record_batches, iox_meta) = match compact_persisting_batch(
+            Arc::new(SystemProvider::new()),
+            &self.exec,
```
Comment (NGA-TRAN) [inline ingester/src/data.rs:213]: 👍

File: influxdb_iox/src/commands/run/ingester.rs [hunk 11]
```diff
@@ -134,6 +135,14 @@ pub struct Config {
         default_value = "1800"
     )]
     pub persist_partition_age_threshold_seconds: u64,
+
+    /// Number of threads to use for the ingester query execution and compaction and persistence.
```
Comment (e-dard) [inline influxdb_iox/src/commands/run/ingester.rs:139]:
```suggestion
    /// Number of threads to use for the ingester query execution, compaction and persistence.
```

File: ingester/src/handler.rs [hunk 12]
```diff
@@ -226,13 +238,21 @@ async fn stream_in_sequenced_entries(
...
         match result {
-            Ok(_) => {
+            Ok(should_pause) => {
                 ingest_recorder.success();
                 span_recorder.ok("stored write");
+
+                if should_pause {
+                    warn!(%sequencer_id, "pausing ingest until persistence has run");
+                    while !lifecycle_manager.can_resume_ingest() {
+                        tokio::time::sleep(Duration::from_millis(100)).await;
```
Comment (e-dard) [inline ingester/src/handler.rs:252]: Is it worth making this a documented constant up the top?

File: ingester/src/lifecycle.rs [hunk 13]
```diff
@@ -142,6 +143,7 @@ impl LifecycleManager {
     /// The persist operations are spawned in new tasks and run at the same time, but the
     /// function waits for all to return before completing.
     pub async fn maybe_persist<P: Persister>(&self, persister: &Arc<P>) {
+        info!("checking persist...");
```
Comment (e-dard) [inline ingester/src/lifecycle.rs:146]: What value is this log line providing? Also, it looks like it can block below this so it seems plausible that seeing this log line doesn't tell you if it's doing any checking at all :-)

What about moving it closer to where something is happening and perhaps adding some useful context to the log line, or alternatively making it `debug` or `trace`?

File: ingester/src/lifecycle.rs [hunk 14]
```diff
@@ -199,6 +201,7 @@ impl LifecycleManager {
             let mut to_persist: Vec<_> = to_persist
                 .into_iter()
                 .map(|s| {
+                    info!("persisting {}", s.partition_id);
```
Comment (e-dard) [inline ingester/src/lifecycle.rs:204]: I wonder if it's worth moving this line into the `persister.persist` function, otherwise you don't know if any persistence is actually happening yet given it's spawned.

File: ingester/src/data.rs [hunk 15]
```diff
@@ -79,48 +93,197 @@ pub struct IngesterData {
...
 impl Persister for IngesterData {
-    async fn persist(&self, _partition_id: PartitionId) {
-        // lookup the TableData
-        // let persisting_batch = table_data.create_persisting_batch(partition.partition_key);
-        // do the persist with this persisting batch
-        // update the catalog
-        // table_data.clear_persisting_batch() (behind the scenes this will remove the persisting batch
-        // and if the partition is empty, remove it from the map in table_data)
+    async fn persist(&self, partition_id: PartitionId) {
+        let retry_time = Duration::from_secs(1);
```
Comment (e-dard) [inline ingester/src/data.rs:143]: Worth making this a constant somewhere?

File: ingester/src/data.rs [hunk 16]
```diff
@@ -511,6 +739,47 @@ impl DataBuffer {
...
+
+        self.snapshot()
+            .expect("This mutable batch snapshot error should be impossible.");
+
+        let mut data = vec![];
+        std::mem::swap(&mut data, &mut self.snapshots);
+        let mut deletes = vec![];
+        std::mem::swap(&mut deletes, &mut self.deletes);
+
+        let queryable_batch = QueryableBatch::new(table_name, data, deletes);
```
Comment (NGA-TRAN) [inline ingester/src/data.rs:768]: Right now, all the deletes are applied to all snapshots which will be wrong for deletes arrived before the data is ingested in the buffer. I think when we do #3699, **everytime a delete comes, we need to snapshot the buffer and add the delete into that snapshot, plus add that delete into the available snapshots, plus keep that delete in the current `deletes` field if there is a persisting is happening**. When we persist, we have to convert each SnapshotBatch into a QueryableBatch, compact each of them, and then merge them into one RecordBatch before saving them into ParquetFile. So there are changes in a few places. I will add this comment into #3699

**Review by NGA-TRAN — COMMENTED** [review]
LGTM. I have discovered a wrong result with delete but it is not the problem of this PR

File: ingester/src/data.rs [hunk 17]
```diff
@@ -79,48 +93,197 @@ pub struct IngesterData {
...
+        let retry_time = Duration::from_secs(1);
+
+        let mut repos = self.catalog.repositories().await;
+
+        // lookup the partition_info from the catalog
+        let partition_info: Option<PartitionInfo> = loop {
+            match repos.partitions().partition_info_by_id(partition_id).await {
+                Ok(p) => break p,
+                Err(e) => {
+                    error!(%e, "error getting partition_info_by_id");
```
Comment (e-dard) [inline ingester/src/data.rs:152]: Should this be logged as an error? An error-level log should be reserved for when something significant has happened that's unrecoverable and an operator need to intervene somehow. But given there is a retry here I don't think that's warranted. What are the circumstances under which this lookup fails? Perhaps `warn` or even `info` might be a better level. 

```suggestion
                    warn!(%e, ?partition_id, "persist partition ID lookup failed");
```

File: ingester/src/data.rs [hunk 18]
```diff
@@ -79,48 +93,197 @@ pub struct IngesterData {
...
-        // and if the partition is empty, remove it from the map in table_data)
+    async fn persist(&self, partition_id: PartitionId) {
+        let retry_time = Duration::from_secs(1);
+
+        let mut repos = self.catalog.repositories().await;
+
+        // lookup the partition_info from the catalog
+        let partition_info: Option<PartitionInfo> = loop {
+            match repos.partitions().partition_info_by_id(partition_id).await {
+                Ok(p) => break p,
```
Comment (e-dard) [inline ingester/src/data.rs:150]: TIL https://doc.rust-lang.org/rust-by-example/flow_control/loop/return.html

File: ingester/src/data.rs [hunk 19]
```diff
@@ -79,48 +93,197 @@ pub struct IngesterData {
...
+                return;
+            }
+        };
+
+        // save the compacted data to a parquet file in object storage
+        loop {
+            match persist(&iox_meta, record_batches.to_vec(), &self.object_store).await {
+                Ok(_) => break,
+                Err(e) => {
+                    error!(%e, "error persisting parquet file to object store");
```
Comment (e-dard) [inline ingester/src/data.rs:239]: I assume this behaviour covers things like networking issues or other transient issues in the CSP? I don't think `error` is warranted here.

```suggestion
                    warn!(%e, "persisting to object store failed: retrying.");
```

**Review by e-dard — COMMENTED** [review]
A few suggestions but generally pretty cool 👍

File: ingester/src/lifecycle.rs (hunk 13 above)
Comment (pauldix) [reply ingester/src/lifecycle.rs:146]: Ah, that was left over debugging, will remove.

File: ingester/src/data.rs (hunk 9 above)
Comment (pauldix) [reply ingester/src/data.rs:250]: Yeah, I'm going to move it inside the function when I update to actually keep track of the min_unpersisted_number in #3702

**Comment by pauldix** [comment]
Ok, addressed @e-dard's comments. Going to move this forward since I think @NGA-TRAN might need some of the bits. Will address any follow on stuff in other PRs and I have #3705 to track a bunch of additional work.
